extensions/opencv/MotionDetector.cpp

```diff
@@ -0,0 +1,222 @@
+        ->isRequired(true)
+        ->withAllowableValues<std::string>({".jpg", ".png"})
+        ->withDefaultValue(".jpg")->build());
+core::Property MotionDetector::MinInterestArea(
+    core::PropertyBuilder::createProperty("Minimum Area")
+        ->withDescription("We only consider the movement regions with area greater than this.")
+        ->isRequired(true)
+        ->withDefaultValue<uint32_t>(100)->build());
+core::Property MotionDetector::Threshold(
+    core::PropertyBuilder::createProperty("Threshold for segmentation")
+        ->withDescription("Pixel greater than this will be white, otherwise black.")
+        ->isRequired(true)
+        ->withDefaultValue<uint32_t>(42)->build());
+core::Property MotionDetector::BackgroundFrame(
+    core::PropertyBuilder::createProperty("Path to background frame")
+        ->withDescription("If not provided then the processor will take the first input frame as background")->build());
+core::Property MotionDetector::DilateIter(
+    core::PropertyBuilder::createProperty("Dilate iteration")
+        ->withDescription("For image processing")
+        ->isRequired(true)
+        ->withDefaultValue<uint32_t>(10)->build());
+core::Property MotionDetector::PrevAsBackground(
+    core::PropertyBuilder::createProperty("Previous frame as Background")
+        ->withDescription("Whether or not detect motion between 2 consecutive frames (yes/no)")
+        ->isRequired(false)
+        ->withAllowableValues<std::string>({"yes", "no"})
+        ->withDefaultValue("no")->build());
+
+core::Relationship MotionDetector::Success(  // NOLINT
+    "success",
+    "Successful to detect motion");
+core::Relationship MotionDetector::Failure(  // NOLINT
+    "failure",
+    "Failure to detect motion");
+
+void MotionDetector::initialize() {
+  std::set<core::Property> properties;
+  properties.insert(ImageEncoding);
+  properties.insert(MinInterestArea);
+  properties.insert(Threshold);
+  properties.insert(BackgroundFrame);
+  properties.insert(DilateIter);
+  properties.insert(PrevAsBackground);
+  setSupportedProperties(std::move(properties));
+
+  setSupportedRelationships({Success, Failure});
+}
+
+void MotionDetector::onSchedule(core::ProcessContext *context, core::ProcessSessionFactory *sessionFactory) {
+  std::string value;
+
+  if (context->getProperty(ImageEncoding.getName(), value)) {
+    image_encoding_ = value;
+  }
+
+  if (context->getProperty(MinInterestArea.getName(), value)) {
+    core::Property::StringToInt(value, min_area_);
+  }
+
+  if (context->getProperty(Threshold.getName(), value)) {
+    core::Property::StringToInt(value, threshold_);
+  }
+
+  if (context->getProperty(DilateIter.getName(), value)) {
+    core::Property::StringToInt(value, dil_iter_);
+  }
+
+  if (context->getProperty(BackgroundFrame.getName(), value) && !value.empty()) {
+    bg_img_ = cv::imread(value, cv::IMREAD_GRAYSCALE);
+    double scale = 500.0 / bg_img_.size().width;
```

Comment:
Is this a value good for all scenarios we can imagine or shall we make it configurable?

Reply:
IMO, this should work for all scenarios, the main point is balance between computation time and accuracy